### PR TITLE
Imviz astrowidgets API MVP triple-play: stretch, cuts, colormap

### LINKED ISSUE #716: Imviz astrowidgets API MVP: stretch_options, stretch
**Reporter:** @pllim 

Implement `stretch_options` and `stretch`. Broken off #631, also see #632.

[🐱](https://jira.stsci.edu/browse/JDAT-1592)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #717: Imviz astrowidgets API MVP: autocut_options, cuts
**Reporter:** @pllim 

Implement `autocut_options` and `cuts`. Broken off #631, also see #632.

[🐱](https://jira.stsci.edu/browse/JDAT-1593)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### LINKED ISSUE #718: Imviz astrowidgets API MVP: colormap_options, set_colormap
**Reporter:** @pllim 

Implement `colormap_options` and `set_colormap`. Broken off #631, also see #632.

[🐱](https://jira.stsci.edu/browse/JDAT-1594)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to implement these astrowidgets API:

* Fix #716 
* Fix #717 
* Fix #718

xref #632 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/imviz/helper.py [hunk 1]
```diff
@@ -286,6 +287,130 @@ def zoom(self, val):
...
+        return viewer.state.layers[i_top].stretch
+
+    @stretch.setter
+    def stretch(self, val):
+        valid_vals = self.stretch_options
+
+        if isinstance(val, type):  # is a class
+            # Translate astropy.visualization
+            from astropy.visualization import AsinhStretch, LinearStretch, LogStretch, SqrtStretch
+            if issubclass(val, AsinhStretch):
```
Comment (duytnguyendtn) [inline jdaviz/configs/imviz/helper.py:352]: Would it be possible to allow the user to put in a string value (i.e. 'log', 'linear', 'asinh', 'sqrt') and have us "autocorrect" behind the scenes to the right `astropy.visualization` class? I feel like that would be a simple `ease-of-use` addon.

Comment (duytnguyendtn) [reply jdaviz/configs/imviz/helper.py:352]: D.U.E. ... nevermind

**Review by duytnguyendtn — APPROVED** [review]
Everything works as expected!

File: jdaviz/configs/imviz/helper.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/imviz/helper.py:352]: Actually, it is the other way around. Glue takes string and we're trying to translate astropy object to the string that Glue understands, as @eteq wants some astropy object support.

Re: D.U.E... I can't tell if you are trying to work your name into the software or I am missing some new CS terminologies here. 😸

Comment (duytnguyendtn) [reply jdaviz/configs/imviz/helper.py:352]: Sorry. "Dumb User Error." I was a dummy. Made a mistake. An oopsie. Please ignore my incompetence ;)

**Comment by ojustino** [comment]
I tried this one with a different set of images and the added functionality worked as intended.

Setting these options triggers some log messages on divide-by-zero errors that I don't see in `main` using the previous `viewer.state.layers` construction. It's from `bqplot` trying to figure out the viewer's aspect ratio, but I think we discussed elsewhere that it's an unrelated issue and it doesn't prevent further interaction with the viewer.

If I blink repeatedly after setting the stretch/cuts/colormap, those options are reset to their default values. This also happens when setting them the old way in `main`. Is that the intended behavior?

**Comment by pllim** [comment]
> If I blink repeatedly after setting the stretch/cuts/colormap, those options are reset to their default values

Sounds like a bug, but I also cannot reproduce it using the example notebook. If you can reproduce this, please tell me how. Thanks!

**Comment by pllim** [comment]
> divide-by-zero errors

In the example notebook, the first cell silences all the warnings, so I don't see them. Handling warnings is a long standing technical debt that we are not able to resolve in time for MVP. 😬

**Comment by ojustino** [comment]
I was imprecise with the wording -- it's not blinking by pressing `B`, it's unchecking and rechecking the image layers in the "Data" menu.

The steps are to run the example notebook from the top through `imviz.cuts = (0, 1000)` and then uncheck and recheck the layers a few times. (The order of unchecking/rechecking doesn't seem to matter.) For me, once the layers are visible again, the stretch/cuts/colormap have all reset to their default values.

**Comment by pllim** [comment]
> uncheck and recheck the layers

Ah, yes. That is a known and unrelated issue, though I am not sure if we have any ticket to address it. Maybe @duytnguyendtn remembers...?

**Comment by duytnguyendtn** [comment]
A quick check on both our JIRA side and repo side for `layers` doesn't yield anything I think is representative of @ojustino's issue. Could you or @pllim make one?

**Comment by pllim** [comment]
Okay, I created #800 for the bug reported in https://github.com/spacetelescope/jdaviz/pull/764#issuecomment-899809890 .

**Review by ojustino — APPROVED** [review]
The core functionality works and both of my concerns were out of scope, so I think this is merge-ready.

**Pull request merged**